multitenant: misc fixes related to tenant capabilities

knz opened the pull request:
See individual commits for details.

The last commit in particular probably addresses #99087.

Epic: CRDB-23559

knz committed:
tenantcapbilitiesauthorizer: the noop authz is exempt from rate limiting

The NoopAuthorizer is meant to pretend that the tenant is not limited
in any way. Previous to this patch, it was still subject to rate
limiting; this patch fixes it.

Release note: None

knz committed:
tenantcapabilitiesauthorizer: avoid using the word "noop"

Prior to this patch, an implementation of the `Authorizer` interface
was called "Noop". Unfortunately, "no-op" does not tell the reader
clearly what an *authorization policy* does. What does it allow? What
does it reject?

This patch clarifies that situation by renaming it to
"AllowEverythingAuthorizer".

Additionally, the RPC context for tenant servers was configured
with this authorizer implementation, which is conceptually confusing:
it would conceptually allow RPC clients to solicit KV operations from
it. Instead, this patch introduces a new "AllowNothingAuthorizer" for
that case.

(Note: this is purely conceptual since the RPC service for tenant
server does not, in fact, offer the roachpb.Internal service.)

Release note: None

knz committed:
kvserver: don't default to "no cap check" authorizer

Prior to this patch, `NewStore()` was tolerant to a missing authorizer
and would synthetize a "allow everything" authorizer in that case.

This is dangerous, because it means any programming error in the
caller where someone forgets to configure the authorizer, would result
in no rate limiting, without any indication something is amiss.

This commit rectifies the situation by requiring an authorizer to be
configured in all cases.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101907)
<!-- Reviewable:end -->

knz committed:
tenantcapabilitiesauthorizer: add a missing test

We want to ensure all KV method types have an entry in the map. This
commit adds a test for it, and adds the missing entry for `WriteBatch`
that was detected by the test.

Release note: None

knz committed:
tenantcapabilitiesauthorizer: fix the fallback setting

Prior to this patch, there was a "authorizer enabled" cluster setting
with just two values on/off.

When off, it would let through all requests.

This was inadequate because for CC Serverless, we want an escape hatch
that disables capability checks, but still disallows sensitive
requests -- i.e. we want a way to restore the pre-v23.1 behavior.

This change achieves this by replacing the previous setting with a new
setting `server.secondary_tenants.authorizer.mode` with 3 values:

- `on`: use capability checks.
- `allow-all`: allow all operations. This can be used e.g. as
  a escape hatch when lifting a single-tenant application into
  multi-tenancy.
- `v222`: restore the pre-v23.1 behavior. In this mode, non-privileged
  operations are allowed for all tenants, and privileged operations are
  only allowed for the system tenant.

Release note: None

knz committed:
tenantcapabilitiesauthorizer: don't use caps until cluster is upgraded

If the cluster version hasn't been upgraded to when capabilities were
introduced, it's better not to start looking at capabilities just yet;
because otherwise error messages would be confusing.

Release note: None

knz committed:
tenantcapabilitiesauthorizer: simplify the fallback

Prior to this patch, we were repeating the following cascade of checks
in each authorizer method:

1. check the cluster setting
2. check whether the reader is bound already
3. check whether the reader has data for the tenant already

Repeating all these checks every time is error prone. Instead, this
change groups them under `getMode`.

Release note: None

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r1, 9 of 9 files at r2, 6 of 6 files at r3, 3 of 3 files at r4, 4 of 4 files at r5, 17 of 18 files at r6, 9 of 9 files at r7, 6 of 6 files at r8, 3 of 3 files at r9, 4 of 4 files at r10, 1 of 1 files at r11, 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101907#-:-NTUsOhO2_Pqdl7eWq9k:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`pkg/kv/kvserver/tenantrate/factory.go` line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101907#-NTUf6fP7D_2r23ljMMo:-NTUf6fP7D_2r23ljMMp:b-sq039d) ([raw file](https://github.com/cockroachdb/cockroach/blob/435df4e6323f97eec20e0038eb46b7413fe5edef/pkg/kv/kvserver/tenantrate/factory.go#L29)):*
> ```Go
> type TestingKnobs struct {
> 	TimeSource timeutil.TimeSource
> 	Authorizer tenantcapabilities.Authorizer
> ```

nit: consider adding a comment here that calls out we don't use this knob if the RPCContext is set. Or, alternatively, we could call this `OverrideAuthorizer` and use it regardless of whether an RPCContext exists or not. This isn't meaningful, but the latter might be more intuitive. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101907#-NTUgAvG7BR0hF6UFzWN:-NTUgAvG7BR0hF6UFzWO:b-vx949j) ([raw file](https://github.com/cockroachdb/cockroach/blob/17d130b1fda61006f722314df323a99a3ef21247/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L106)):*
> ```Go
> 			// All allowable request types must be explicitly opted into the
> 			// reqMethodToCap map. If a request type is missing from the map
> 			// (!hasCap), we must be conservative and assume it is
> ```

Now that we've added this test, should we pull out the `!hasCap` from here and make a stronger assertion about it instead? Maybe a Fatal?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 334 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/101907#-NTUs8SoDSOxWASS1SQe:-NTUs8SoDSOxWASS1SQf:b9fgczf) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b689c3b1f4899653f789137df3f540efd03c53/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L334)):*
> ```Go
> 		return false
> 	default:
> 		log.Warningf(ctx, "unknown authorizer mode: %d", mode)
> ```

Should this, and in other places, use a Fatal?


<!-- Sent from Reviewable.io -->

knz committed:
tenantcapabilitiesauthorizer: use `ReportOrPanic` where applicable

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101907)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvserver/tenantrate/factory.go` line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101907#-NTUf6fP7D_2r23ljMMo:-NTVNqCv7exEsNdExOzs:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/435df4e6323f97eec20e0038eb46b7413fe5edef/pkg/kv/kvserver/tenantrate/factory.go#L29)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider adding a comment here that calls out we don't use this knob if the RPCContext is set. Or, alternatively, we could call this `OverrideAuthorizer` and use it regardless of whether an RPCContext exists or not. This isn't meaningful, but the latter might be more intuitive. 
</blockquote></details>

Good idea. Done.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101907#-NTUgAvG7BR0hF6UFzWN:-NTVNvYG5UxazD_vPEyI:bm3w79t) ([raw file](https://github.com/cockroachdb/cockroach/blob/17d130b1fda61006f722314df323a99a3ef21247/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L106)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Now that we've added this test, should we pull out the `!hasCap` from here and make a stronger assertion about it instead? Maybe a Fatal?
</blockquote></details>

nah; I can see someone modifying the map, forgetting to run the test, then running their server with the binary. This is a good defense.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 334 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/101907#-NTUs8SoDSOxWASS1SQe:-NTVPLL_0ie_lCL-k9QG:bp2pw60) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b689c3b1f4899653f789137df3f540efd03c53/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L334)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this, and in other places, use a Fatal?
</blockquote></details>

`log.Fatal` is _almost always_ code smell and should be generally avoided (and so should be `panic`).

But this is probably a good use case for `ReportOrPanic`. Added a commit to do that.


<!-- Sent from Reviewable.io -->

knz commented:
bors r=arulajmani
